### PR TITLE
refactored l1_loss and margin_ranking_loss in paddle frontend 

### PR DESCRIPTION
**Pull request by zeus2x7**
refactored l1_loss,nll_loss and margin_ranking_loss in paddle frontend

### CODE REVIEW

**Comment by zeus2x7** [comment]
hi @lucasalavapena  can you please review the pr?

**Comment by zeus2x7** [comment]
hi @lucasalavapena  it would be much helpful if you could review the pr

**Comment by lucasalavapena** [comment]
Hi @zeus2x7 . Thanks for your PR :smile: . Sorry for the wait, it was the weekend.

Could you please split your PR into 2? It is the commended approach per https://unify.ai/docs/ivy/overview/contributing/the_basics.html#small-commits-often . Ideally you could make one PR for the refactor and one into adding the normal function? You can assign me to both if you want as I will anyways now comment on both.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py [hunk 1]
```diff
@@ -1,12 +1,52 @@
...
 
 # local
 
 import ivy_tests.test_ivy.helpers as helpers
 from ivy_tests.test_ivy.helpers import handle_frontend_test
 
 
+@handle_frontend_test(
+    fn_tree="paddle.normal",
+    input_dtypes=st.sampled_from([["int32"], ["int64"]]),
```
Comment (lucasalavapena) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py:13]: Why are you sampling from ints? If you support floats?

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py [hunk 2]
```diff
@@ -1,12 +1,52 @@
 # global
 from hypothesis import strategies as st
+import ivy
```
Comment (lucasalavapena) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py:3]: I think if you change how the argument of `input_dtypes` in `handle_frontend_test`  as I mentioned below, you can get rid of this import

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py [hunk 3]
```diff
@@ -1,12 +1,52 @@
...
+    shape,
+    mean,
+    std,
+    frontend,
+    backend_fw,
+    test_flags,
+    fn_tree,
+):
+    helpers.test_frontend_function(
+        input_dtypes=input_dtypes + [ivy.float64] + [ivy.float32],
```
Comment (lucasalavapena) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_random.py:38]: It probably makes sense to handle the types in  argument of `input_dtypes` in `handle_frontend_test`.

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -274,6 +272,29 @@ def nll_loss(
     """Refer
     https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss for
     more on NLL(Negative log likelihood) Loss."""
+    if reduction not in ["sum", "mean", "none"]:
+        raise ValueError(
```
Comment (lucasalavapena) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:276]: For all the value errors I would sugges to  `ivy.utils.exceptions.IvyValueError` see [0, 1].

[0] https://unify.ai/docs/ivy/docs/utils/ivy.utils.exceptions.html#ivy.utils.exceptions.IvyValueError
[1] https://unify.ai/docs/ivy/overview/deep_dive/exception_handling.html

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by zeus2x7** [comment]
hi @lucasalavapena  can you please review the pr again?
have made all the changes suggested by you

**Comment by zeus2x7** [comment]
@lucasalavapena  could you please review this pr also

**Comment by lucasalavapena** [comment]
Please have some patience.

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -287,11 +309,22 @@ def nll_loss(
...
-        return output
-    num = ivy.sum(loss)
-    output = num / den
-    if ignore_index >= 0 and ignore_index < ivy.shape(input)[1]:
-        output = output - loss[ignore_index] / den
+
+    elif reduction == "mean":
+        output = ivy.mean(loss)
+        if ignore_index >= 0 and ignore_index < ivy.shape(input)[1]:
+            output = output - loss[ignore_index]
```
Comment (lucasalavapena) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:316]: This is not a mean, mean without index != mean - value

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 6]
```diff
@@ -287,11 +309,22 @@ def nll_loss(
...
+
+    elif reduction == "none":
+        output = loss
+        if ignore_index >= 0 and ignore_index < ivy.shape(input)[1]:
+            output = output - loss[ignore_index] / den
```
Comment (lucasalavapena) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:321]: Could you explain your reasoning here? This is not what I would have expected here

**Comment by lucasalavapena** [comment]
Btw a general comment, something like nll_loss should ideally be calling the ivy functional api, but that function has not been implemented yet and of course you are simply completing/refactoring it!

**Comment by zeus2x7** [comment]
hi @lucasalavapena  can you please give it a look now?

**Review by lucasalavapena — COMMENTED** [review]
Please ensure that all tests related to this PR pass, `test_paddle_nll_loss` is failing (see the combined test results section of display test results). I first realised this when I ran it in codespaces because I was confused how it would be passing given that log or log_softmax is applied here.
